In [8]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [5]:
os.getcwd()

'C:\\Users\\yasho'

In [7]:
os.chdir("F:\PYTHON\python statistics and ML classes notes\Machine Learning Classes\Supevice Machine Learning\projects in jupyter")

In [9]:
dataset = pd.read_csv('50_Startups.csv')
dataset.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

# Pre-processing
# part 1 : handling missing value
# part 2 : handling encoding part (label encoder, one hot encoder, dummy)
# part 3 : handling outlier - transaformation concept / regularisation 
# part 4 : feature scaling - standarisation and normalisation
# part 5 : handling imbalance dataset

In [10]:
# part 1 : check missing data
dataset.isnull().sum()

R&D Spend          0
Administration     0
Marketing Spend    0
State              0
Profit             0
dtype: int64

In [11]:
# part 2 : Encoding 
# Label encoder - changing category to codes
dataset['State'] = dataset['State'].astype('category')
dataset['State'] = dataset['State'].cat.codes

In [12]:
dataset['State'].value_counts()

2    17
0    17
1    16
Name: State, dtype: int64

In [13]:
dataset.head()

R&D Spend  Administration  Marketing Spend  State     Profit
0  165349.20       136897.80        471784.10      2  192261.83
1  162597.70       151377.59        443898.53      0  191792.06
2  153441.51       101145.55        407934.54      1  191050.39
3  144372.41       118671.85        383199.62      2  182901.99
4  142107.34        91391.77        366168.42      1  166187.94

In [18]:
# One Hot Encoder
dataset = pd.get_dummies(dataset, columns=['State'])

In [19]:
dataset.head()

R&D Spend  Administration  Marketing Spend     Profit  State_0  State_1  \
0  165349.20       136897.80        471784.10  192261.83        0        0   
1  162597.70       151377.59        443898.53  191792.06        1        0   
2  153441.51       101145.55        407934.54  191050.39        0        1   
3  144372.41       118671.85        383199.62  182901.99        0        0   
4  142107.34        91391.77        366168.42  166187.94        0        1   

   State_2  
0        1  
1        0  
2        0  
3        1  
4        0

In [20]:
dataset = dataset.drop(['State_0'], axis=1)

In [21]:
dataset.head()

R&D Spend  Administration  Marketing Spend     Profit  State_1  State_2
0  165349.20       136897.80        471784.10  192261.83        0        1
1  162597.70       151377.59        443898.53  191792.06        0        0
2  153441.51       101145.55        407934.54  191050.39        1        0
3  144372.41       118671.85        383199.62  182901.99        0        1
4  142107.34        91391.77        366168.42  166187.94        1        0

In [22]:
#spliting the variables into x and y.....
x = dataset.drop(['Profit'],axis=1)
y = dataset['Profit']

print(x.head())
print('\n')
print(y.head())

   R&D Spend  Administration  Marketing Spend  State_1  State_2
0  165349.20       136897.80        471784.10        0        1
1  162597.70       151377.59        443898.53        0        0
2  153441.51       101145.55        407934.54        1        0
3  144372.41       118671.85        383199.62        0        1
4  142107.34        91391.77        366168.42        1        0


0    192261.83
1    191792.06
2    191050.39
3    182901.99
4    166187.94
Name: Profit, dtype: float64


In [23]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,train_size=0.75,random_state=101)

print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(37, 5)
(13, 5)
(37,)
(13,)


# Decision Tree Regressor

In [24]:
from sklearn.tree import DecisionTreeRegressor
dt_model = DecisionTreeRegressor()
dt_model.fit(x_train, y_train)

DecisionTreeRegressor()

In [25]:
y_pred_train = dt_model.predict(x_train)
y_pred_test = dt_model.predict(x_test)

In [26]:
from sklearn.metrics import r2_score

print(r2_score(y_train, y_pred_train))
print("\n")
print(r2_score(y_test, y_pred_test))


1.0


0.903160139515659


# RandomForest Regressor

In [28]:
from sklearn.ensemble import RandomForestRegressor
rf_model = RandomForestRegressor()
rf_model.fit(x_train,y_train)

RandomForestRegressor()

In [29]:
y_pred_rf_train = rf_model.predict(x_train)
y_pred_rf_test = rf_model.predict(x_test)

In [30]:
print(r2_score(y_train, y_pred_rf_train))
print("\n")
print(r2_score(y_test, y_pred_rf_test))

0.9849379448229224


0.9564298963217992


# Linear Regression

In [31]:
from sklearn.linear_model import LinearRegression
lr_model = LinearRegression()
lr_model.fit(x_train, y_train)

LinearRegression()

In [32]:
y_pred_lr_train = lr_model.predict(x_train)
y_pred_lr_test = lr_model.predict(x_test)

In [33]:
print(r2_score(y_train, y_pred_lr_train))
print("\n")
print(r2_score(y_test, y_pred_lr_test ))

0.9473736604055617


0.9428740924766765


# conclude :

### Decision Tree Regressor : Accuracy ****************
##### Train : 100%
##### Test : 88.05%


### Random Forest Regressor : Accuracy
##### Train : 99%
##### Test : 84.5%

### Linear Regression : Accuracy
##### Train : 96%
##### Test : 82.4%